src/codegen/compile.ml

```diff
@@ -6602,6 +6848,25 @@ module MakeSerialization (Strm : Stream) = struct
           Tagged.allocation_barrier env ^^
           G.i Drop
         )
+      | Prim Region ->
+         read_alias env (Prim Region) (fun get_region_typ on_alloc ->
+          let (set_region, get_region) = new_local env "region" in
+          (* sanity check *)
+          get_region_typ ^^
+          compile_eq_const (Int32.neg (Option.get (to_idl_prim (Prim Region)))) ^^
+          E.else_trap_with env "deserialize_go (Region): unexpected idl_typ" ^^
+          Region.alloc env (compile_unboxed_const 0l) (compile_unboxed_const 0l) (Blob.lit env "") ^^ set_region ^^
+          on_alloc get_region ^^
+          get_region ^^ ReadBuf.read_word32 env get_data_buf ^^ Region.store_field env Region.id_field ^^
+          get_region ^^ ReadBuf.read_word32 env get_data_buf ^^ Region.store_field env Region.page_count_field ^^
+          (*@Luc: vec_pages_field object field update! do we need a write_barrier or is the allocation barrier enough?
+            Note that Region.alloc above already containes an allocation_barrier, but we later update the vec_pages_field.
+           *)
+          get_region ^^ read_blob () ^^ Region.store_field env Region.vec_pages_field ^^
+          get_region ^^
+          Tagged.allocation_barrier env ^^
+          G.i Drop
```

Comment:
@luc-blaeser can you check this code is correct regarding barriers? I added what I believe to be a missing one in 2fd28e48654ae46437670077989b7f618ce7755f

Reply:
Maybe this should just be a write_barrier, since the allocation already contains an allocation_barrier?

Reply:
Also related: https://github.com/dfinity/motoko/pull/3768#discussion_r1270729262
(Somehow got outdated by GitHub...)

Reply:
Done